### PR TITLE
RT 1.1 base_bgp_session_parameters_test.go

### PR DESCRIPTION
**Pull request by cprabha**
Hi,

Md5 password configuration and verification is moved outside of confirm.state() as password will be in encrypted state.

Thanks,
Prabha

### CODE REVIEW

**Review by sachendras — APPROVED** [review]
LGTM.

For Google folks, complete internal assessment covered in b/260787822

**Comment by xw-g** [comment]
While we take time to debate how should the OC handles BGP password, for this test I think we can remove the password validation steps? We can rely on the BGP establishment to validate if the right password is given/configured.

**Comment by cprabha** [comment]
> While we take time to debate how should the OC handles BGP password, for this test I think we can remove the password validation steps? We can rely on the BGP establishment to validate if the right password is given/configured.

@xw-g  I have commented out password checks , please could you  review and merge the code.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 1]
```diff
@@ -349,7 +368,6 @@ func TestParameters(t *testing.T) {
 	}
 	for _, tc := range cases {
 		t.Run(tc.name, func(t *testing.T) {
-			fptest.LogQuery(t, "DUT BGP Config before", dutConfPath.Config(), gnmi.GetConfig(t, dut, dutConfPath.Config()))
```
Comment (xw-g) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:352]: what's the motivation behind removing this log?

**Review by xw-g — COMMENTED** [review]
A few thoughts in general: 
* Can you help update the description to reflect what are expected to be changed via this PR?  Seems like current description is outdated. 
* Since we are now validating the status indirectly, I think we may wanted to add a negative test case? e.g. configure the wrong password on purpose and observe the BGP didn't come up.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 2]
```diff
@@ -279,17 +293,22 @@ func TestEstablishAndDisconnect(t *testing.T) {
 	t.Log("Check BGP parameters")
 	verifyBgpTelemetry(t, dut)
 
+	// Password authentication is currently disabled as how OC handles BGP password
```
Comment (xw-g) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:296]: @liulk Should we remove the code or keep them in comment state? I don't think we will uncomment these code soon. What's the style guidance here? Thanks.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 1 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:352]: @xw-g , If there is no config present on DUT (this step is before configuring test config) it will fail, Hence removing it.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 3]
```diff
@@ -305,7 +324,7 @@ func TestEstablishAndDisconnect(t *testing.T) {
 
 // TestParameters is to verify normal session establishment and termination
 // in both eBGP and iBGP scenarios using session parameters like explicit
-// router id , timers and MD5 authentication.
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:308]: @xw-g , This is moved to TestEstablishAndDisconnect testcase hence removed the comment under TestParameters testcase.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 4]
```diff
@@ -220,17 +231,18 @@ func configureATE(t *testing.T, ateParams *bgpTestParams, connectionType connTyp
...
-			WithMD5Key(authPassword).WithHoldTime(ateHoldTime)
+			WithHoldTime(ateHoldTime)
 	}
 	return topo
 
 }
 
 // TestEstablishAndDisconnect Establishes BGP session between DUT and ATE and Verifies
 // abnormal termination of session using notification message:
+// Also configures and verifies MD5 authentication password
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: added comment saying md5 authentication is implicit via BGP adajcency.

**Comment by xw-g** [comment]
Since we are now validating the status indirectly, I think we may want to add a negative test case? e.g. configure the wrong password on purpose and observe the BGP didn't come up?

**Comment by cprabha** [comment]
> Since we are now validating the status indirectly, I think we may want to add a negative test case? e.g. configure the wrong password on purpose and observe the BGP didn't come up?

@xw-g Sure, can add one more testcase to verify negative scenario.

**Comment by cprabha** [comment]
@xw-g , I have added teststeps to verify negative scenario for password mismatch.
also updated README file.
Please kindly review the changes.

Thanks,
Prabha

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md [hunk 5]
```diff
@@ -16,7 +16,11 @@ Test the abnormal termination of session using notification message:
 
     *   Ensure session state should be `ESTABLISHED`.
     *   Verify BGP capabilities: route refresh, ASN32 and MPBGP.
-
+    *   Enable MD5 authentication on DUT and ATE. Password verification
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:19]: The two added bullet points do not fit well to the flow of the test plan. This section is about establishing a session and terminate it abnormally. The `* Establish BGP session` bullet point (line 15) is about establishing the session, and `* Verify BGP session disconnect` (line 24) is about terminating the session abnormally.

The section below is about establishing and terminating a session normally, so the password mismatch test doesn't really fit there either.

The negative password test is about abnormal session establishment, so it should be in its own section. See suggested edit below.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md [hunk 6]
```diff
@@ -37,7 +41,6 @@ Test the normal session establishment and termination:
     And include the following session parameters for all cases:
 
     *   Explicitly specified Router ID.
-    *   Enable MD5 authentication on DUT and ATE.
     *   Explicit holdtime interval and keepalive interval.
     *   Explicit connect retry interval.
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:46]:
```suggestion

Test the password authentication on session establishment:

*   Configure matching passwords on DUT and ATE. Verify that BGP adjacency succeeds.
*   Configure mismatched passwords on DUT and ATE. Verify that BGP adjacency fails.
```

Comment (liulk) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:46]: This should be done in its own function in test code:

```
func TestPassword(t *testing.T) {
  ...
}
```

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 2 above)
Comment (liulk) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:296]: Since we're not supposed to be able to retrieve the password back from the device, let's remove the code. It's just confusing when left as a comment.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 7]
```diff
@@ -151,6 +149,19 @@ func verifyBGPCapabilities(t *testing.T, dut *ondatra.DUTDevice) {
 
 }
 
+// verifyAuthPassword checks that the dut applied configured auth password to bgp neighbors
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:152]: Remove since we're not supposed to be able to fetch the password back from the device.

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 2 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Removed above mentioned code from script. Thank you.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md [hunk 8]
```diff
@@ -16,7 +16,11 @@ Test the abnormal termination of session using notification message:
 
     *   Ensure session state should be `ESTABLISHED`.
     *   Verify BGP capabilities: route refresh, ASN32 and MPBGP.
-
+    *   Enable MD5 authentication on DUT and ATE. Password verification
+        is implicit, done through bgp adajcency.
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md]: updated, Thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md [hunk 9]
```diff
@@ -16,7 +16,11 @@ Test the abnormal termination of session using notification message:
...
+    *   Configure different password on DUT, Verify if BGP adajcency fails when     
+        passwords are not matching on DUT - ATE.
```
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md]: updated, thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md (hunk 6 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md:49]: updated , thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 7 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Removed this part of code, thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md (hunk 5 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/README.md]: Added TestPassword function, Thank you.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 10]
```diff
@@ -286,10 +286,10 @@ func TestEstablishAndDisconnect(t *testing.T) {
 	// Send Cease Notification from ATE to DUT
 	t.Log("Send Cease Notification from ATE to DUT")
 	ate.Actions().NewBGPPeerNotification().WithCode(6).WithSubCode(6).WithPeers(bgpPeer).Send(t)
-
+	time.Sleep(time.Second * 3)
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:289]: Is this sleep still needed?

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 11]
```diff
@@ -303,9 +303,79 @@ func TestEstablishAndDisconnect(t *testing.T) {
...
+	time.Sleep(time.Second * dutHoldTime)
+
+	t.Log("Verify BGP session state : Should not be in ESTABLISHED state when passwords does not match")
+	status := gnmi.Get(t, dut, nbrPath.SessionState().State())
+	if status == oc.Bgp_Neighbor_SessionState_ESTABLISHED {
+		t.Logf("BGP Adajcency is UP when passwords are not matching")
+		t.Errorf("Md5 authentication verification failed, %v", status)
+	}
+
+	t.Logf("Revert md5 auth password on DUT to match with ATE.")
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:366]:
```suggestion
	t.Log("Revert md5 auth password on DUT to match with ATE.")
```

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 12]
```diff
@@ -303,9 +303,79 @@ func TestEstablishAndDisconnect(t *testing.T) {
...
+
+	t.Logf("Configure mismatching md5 auth password on DUT")
+	gnmi.Replace(t, dut, dutConfPath.Neighbor(ateAttrs.IPv4).AuthPassword().Config(), "PASSWORDNEGSCENARIO")
+	t.Logf("wait till hold time expires...")
+	time.Sleep(time.Second * dutHoldTime)
+
+	t.Log("Verify BGP session state : Should not be in ESTABLISHED state when passwords does not match")
+	status := gnmi.Get(t, dut, nbrPath.SessionState().State())
+	if status == oc.Bgp_Neighbor_SessionState_ESTABLISHED {
+		t.Logf("BGP Adajcency is UP when passwords are not matching")
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:362]:
```suggestion
		t.Log("BGP Adajcency is UP when passwords are not matching")
```

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 13]
```diff
@@ -303,9 +303,79 @@ func TestEstablishAndDisconnect(t *testing.T) {
...
+	gnmi.Delete(t, dut, dutConfPath.Config())
+	dutConf := bgpCreateNbr(&bgpTestParams{localAS: dutAS, peerAS: ateAS})
+	gnmi.Replace(t, dut, dutConfPath.Config(), dutConf)
+	t.Log("Configure matching Md5 auth password on DUT")
+	gnmi.Replace(t, dut, dutConfPath.Neighbor(ateAttrs.IPv4).AuthPassword().Config(), authPassword)
+
+	fptest.LogQuery(t, "DUT BGP Config", dutConfPath.Config(), gnmi.GetConfig(t, dut, dutConfPath.Config()))
+
+	// ATE Configuration.
+	t.Log("configure port and BGP configs on ATE")
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:336]:
```suggestion
	t.Log("Configure port and BGP configs on ATE")
```

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 14]
```diff
@@ -303,9 +303,79 @@ func TestEstablishAndDisconnect(t *testing.T) {
...
+
+	t.Log("Pushing config to ATE and starting protocols...")
+	topo.Push(t)
+	topo.StartProtocols(t)
+
+	// Verify BGP status
+	t.Log("Check BGP parameters")
+	verifyBgpTelemetry(t, dut)
+
+	t.Logf("Configure mismatching md5 auth password on DUT")
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:354]:
```suggestion
	t.Log("Configure mismatching md5 auth password on DUT")
```

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 15]
```diff
@@ -303,9 +303,79 @@ func TestEstablishAndDisconnect(t *testing.T) {
...
+	gnmi.Replace(t, dut, dutConfPath.Neighbor(ateAttrs.IPv4).AuthPassword().Config(), "PASSWORDNEGSCENARIO")
+	t.Logf("wait till hold time expires...")
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:356]:
```suggestion
	t.Log("Wait till hold time expires...")
```

**Review by liulk — CHANGES_REQUESTED** [review]
Mostly LGTM, except for some style fixes.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 15 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Updated, Thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 14 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Updated, Thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 13 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Updated, Thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 12 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Updated, Thank you

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 11 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: Updated, Thank you

**Review by liulk — APPROVED** [review]

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 10 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:289]: Yeah, Since we are sending notification from Ixia , 2-3 secs is needed.

**Comment by xw-g** [comment]
@cprabha can you help confirm if all the affected tests are still green with this PR? Especially, is `TestPassword` green with your testbed? Thanks.

**Comment by xw-g** [comment]
@vcheny @vosipchu @robshakir because this PR is mainly about introducing a new BGP MD5 negative tests `TestPassword`, following the process, I think 1 LGTM (already got from @sachendras ) and 1 code review (already got from @liulk ) is good enough to merge. Let me know if any objection. Thanks.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 16]
```diff
@@ -359,7 +428,8 @@ func TestParameters(t *testing.T) {
 			tc.ateConf.Push(t)
 			tc.ateConf.StartProtocols(t)
 			t.Log("Verify BGP session state : ESTABLISHED")
-			gnmi.Await(t, dut, nbrPath.SessionState().State(), time.Second*100, oc.Bgp_Neighbor_SessionState_ESTABLISHED)
+			time.Sleep(time.Second * 5)
```
Comment (xw-g) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:431]: @liulk since there is already a `gnmi.Await` below, is `time.Sleep()` here needed? Thanks.

**Comment by cprabha** [comment]
> This is a merge blocking comment.

@xw-g , Please check below passed log on our setup.
[bgp_RT1.1_TestPassword.txt](https://github.com/openconfig/featureprofiles/files/10159466/bgp_RT1.1_TestPassword.txt)

Thanks,
Prabha

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 16 above)
Comment (liulk) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:431]: I had a similar comment earlier: https://github.com/openconfig/featureprofiles/pull/787/files/cdc08fb4610aaab9c52eae1d3891de836db45226#r1040170605

I think this is very subtle. If I understand the code correctly, Await will fatal immediately on subscription error which can happen if the path has not become valid.

https://github.com/openconfig/ygnmi/blob/a492ddad11462bfe95df4b6ddda2c0f6453e1168/ygnmi/ygnmi.go#L306-L310

@DanG100 I think the sleep + await is an anti-pattern. Is there a way we can just await with implicit subscription retries?

@xw-g @cprabha I think it's fine to do sleep + await for now.

Comment (wenovus) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:431]: If the path is valid and supported by the target, it MUST not return an error -- rather no data is received. Thus the stream should not be closed, and the sleep is therefore unnecessary if this behaviour is followed.

https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#3524-subscriberesponse-
behavior-table

Comment (liulk) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:431]: nbrPath is keyed by ateIP: `/network-instances/network-instance[name=DEFAULT]/protocols/protocol[identifier=BGP]/bgp/neighbors/neighbor[neighbor-address=ateIP]`

Are you saying that even though ateIP is missing, the path is still syntactically correct, so the device should just return no value?

**Comment by cprabha** [comment]
@liulk , @xw-g  removed sleep timer. Please check changes.

Thanks,
Prabha

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 17]
```diff
@@ -286,10 +286,8 @@ func TestEstablishAndDisconnect(t *testing.T) {
 	// Send Cease Notification from ATE to DUT
 	t.Log("Send Cease Notification from ATE to DUT")
 	ate.Actions().NewBGPPeerNotification().WithCode(6).WithSubCode(6).WithPeers(bgpPeer).Send(t)
-
-	// Verify BGP session state : ACTIVE
 	t.Log("Verify BGP session state : ACTIVE")
-	gnmi.Await(t, dut, nbrPath.SessionState().State(), time.Second*60, oc.Bgp_Neighbor_SessionState_ACTIVE)
+	gnmi.Await(t, dut, nbrPath.SessionState().State(), time.Second*20, oc.Bgp_Neighbor_SessionState_ACTIVE)
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:290]: @xw-g this actually reduced the await time. Are you okay with this?

Comment (xw-g) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:290]: @sachendras to comment further.

@liulk feel free to merge when all your comments are resolved.

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 18]
```diff
@@ -359,7 +426,7 @@ func TestParameters(t *testing.T) {
 			tc.ateConf.Push(t)
 			tc.ateConf.StartProtocols(t)
 			t.Log("Verify BGP session state : ESTABLISHED")
-			gnmi.Await(t, dut, nbrPath.SessionState().State(), time.Second*100, oc.Bgp_Neighbor_SessionState_ESTABLISHED)
+			gnmi.Await(t, dut, nbrPath.SessionState().State(), time.Second*30, oc.Bgp_Neighbor_SessionState_ESTABLISHED)
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:429]: Revert the await timeout reduction here also?

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 17 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: updated time back to 60 secs

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go [hunk 19]
```diff
@@ -349,7 +417,6 @@ func TestParameters(t *testing.T) {
 	}
 	for _, tc := range cases {
 		t.Run(tc.name, func(t *testing.T) {
-			fptest.LogQuery(t, "DUT BGP Config before", dutConfPath.Config(), gnmi.GetConfig(t, dut, dutConfPath.Config()))
```
Comment (liulk) [inline feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go:352]: Revert this change?

**Review by liulk — APPROVED** [review]

File: feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go (hunk 18 above)
Comment (cprabha) [reply feature/experimental/bgp/ate_tests/base_bgp_session_parameters/base_bgp_session_parameters_test.go]: updated, thank you

**Review by sachendras — APPROVED** [review]
